File: Source/App/Scripts/App/Models/Organization.ts
/// <reference path="../exceptionless.ts" />

module exceptionless.models {
    export class Organization {
        id: string;
        name: string;
        projectCount: number;
        stackCount: number;
        errorCount: number;
        totalErrorCount: number;
        lastErrorDate: Date;
        subscribeDate: Date;

        billingChangeDate: Date;
        billingChangedByUserId: string;
        billingStatus = enumerations.BillingStatus.Active;
        billingPrice: number;

        planId: string;
        cardLast4: string;
        stripeCustomerId: string;

        isSuspended: boolean;
        suspensionCode: string;
        suspensionDate: Date;
        suspendedByUserId: string;
        suspensionNotes: string;

        constructor(id: string, name: string, projectCount: number, stackCount: number, errorCount: number, totalErrorCount: number, lastErrorDate?: Date, subscribeDate?: Date, billingChangeDate?: Date, billingChangedByUserId?: string, billingStatus?: number, billingPrice?:number, planId?: string, cardLast4?: string, stripeCustomerId?: string, isSuspended?: boolean, suspensionCode?: string, suspensionDate?: Date, suspendedByUserId?: string, suspensionNotes?: string) {
            this.id = id;
            this.name = name;
            this.projectCount = projectCount;
            this.stackCount = stackCount;
            this.errorCount = errorCount;
            this.totalErrorCount = totalErrorCount;

            this.lastErrorDate = lastErrorDate;

            if (subscribeDate)
                this.subscribeDate = subscribeDate;
            
            if (billingStatus)
                this.billingStatus = <enumerations.BillingStatus>billingStatus;

            if (billingChangeDate)
                this.billingChangeDate = billingChangeDate;

            this.billingChangedByUserId = !StringUtil.isNullOrEmpty(billingChangedByUserId) ? billingChangedByUserId : '';
            this.billingPrice = billingPrice;
            this.planId = planId;
            this.cardLast4 = cardLast4;
            this.stripeCustomerId = stripeCustomerId;

            this.isSuspended = isSuspended == true;
            this.suspensionCode = !StringUtil.isNullOrEmpty(suspensionCode) ? suspensionCode : '';
            this.suspendedByUserId = !StringUtil.isNullOrEmpty(suspendedByUserId) ? suspendedByUserId : '';
            this.suspensionNotes = !StringUtil.isNullOrEmpty(suspensionNotes) ? suspensionNotes : '';

            if (suspensionDate)
                this.suspensionDate = suspensionDate;
            
            ko.track(this);
        }

        public get createdOn(): Date {
            return new Date(parseInt(this.id.substring(0, 8), 16) * 1000);
        }

        public get projects(): models.ProjectInfo[] {
            return ko.utils.arrayFilter(App.projects(), (project: models.ProjectInfo) => {
                return project.organizationId === this.id;
            });
        }

        public get selectedPlan(): account.BillingPlan {
            var plan = ko.utils.arrayFirst(App.plans(), (plan: account.BillingPlan) => plan.id === this.planId);
            return plan ? plan : new account.BillingPlan(Constants.FREE_PLAN_ID, 'Free', 'Free', 0, false, 1, 2500, 5, 1, 7, false);
        }

        public get billingStatusAbbr(): string {
            if (this.isSuspended)
                return '[S]';

            return '[' + enumerations.BillingStatus[this.billingStatus].toUpperCase()[0] + ']';
        }

        public get billingSummary(): string {
            var summary = '<h5>Billing (' + enumerations.BillingStatus[this.billingStatus].toString() + ')</h5>';
            summary += '<p>Subscribed on: <strong>' + DateUtil.format(moment(this.subscribeDate)) + '</strong></p>';

            if (this.billingChangeDate)
                summary += '<p>Last changed: <strong>' + DateUtil.format(moment(this.billingChangeDate)) + '</strong></p>';

            if (!StringUtil.isNullOrEmpty(this.billingChangedByUserId))
                summary += '<p>Changed by: <strong>' + this.billingChangedByUserId + '</strong></p>';

            if (!StringUtil.isNullOrEmpty(this.cardLast4))
                summary += '<p>Card Last 4: <strong>' + this.cardLast4 + '</strong></p>';

            if (this.isSuspended) {
                summary +=  '<h5>Suspension information</h5>';
                summary += '<p>Suspended on: <strong>' + DateUtil.format(moment(this.suspensionDate)) + '</strong></p>';
                summary += '<p>Suspended by: <strong>' + this.suspendedByUserId + '</strong></p>';
                summary += '<p>Reason: <strong>' + this.suspensionCode + '</strong></p>';

                if (!StringUtil.isNullOrEmpty(this.suspensionNotes))
                    summary += '<p>Notes: <strong>' + this.suspensionNotes + '</strong></p>';
            }

            return summary;
        }

        public get retentionSummary(): string {
            var summary = '<p>Projects: <strong>' + numeral(this.projectCount).format('0,0[.]0') + '</strong></p>';
            summary += '<p>Stacks: <strong>' + numeral(this.stackCount).format('0, 0[.]0') + '</strong></p>';
            summary += '<p>Errors: <strong>' + numeral(this.errorCount).format('0, 0[.]0') + '</strong></p>';
            summary += '<p>Total Errors: <strong>' + numeral(this.totalErrorCount).format('0, 0[.]0') + '</strong></p>';
            return summary;
        }

        public get activitySummary(): string {
            var activity = '<p>Created on: <strong>' + DateUtil.format(moment(this.createdOn)) + '</strong></p>';
            
            if (this.subscribeDate)
                activity += '<p>Subscribed on: <strong>' + DateUtil.format(moment(this.subscribeDate));
            
            var lastActivity = this.lastErrorDate && <any>this.lastErrorDate !== "0001-01-01T00:00:00Z" ? DateUtil.format(moment(this.lastErrorDate)) : 'Never';
            activity += '</strong></p><p>Last Error: <strong> ' + lastActivity + '</strong></p>';
            
            return activity;
        }
    }
}

File: Source/Models/Admin/Organization.cs
#region Copyright 2014 Exceptionless

// This program is free software: you can redistribute it and/or modify it 
// under the terms of the GNU Affero General Public License as published 
// by the Free Software Foundation, either version 3 of the License, or 
// (at your option) any later version.
// 
//     http://www.gnu.org/licenses/agpl-3.0.html

#endregion

using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;

namespace Exceptionless.Models {
    public class Organization : IIdentity {
        public Organization() {
            Invites = new Collection<Invite>();
            BillingStatus = BillingStatus.Trialing;
            OverageHours = new Collection<UsageInfo>();
            Usage = new Collection<UsageInfo>();
        }

        /// <summary>
        /// Unique id that identifies the organization.
        /// </summary>
        public string Id { get; set; }

        /// <summary>
        /// Name of the organization.
        /// </summary>
        public string Name { get; set; }

        /// <summary>
        /// Stripe customer id that will be charged.
        /// </summary>
        public string StripeCustomerId { get; set; }

        /// <summary>
        /// Billing plan that the organization belongs to.
        /// </summary>
        public string PlanId { get; set; }

        /// <summary>
        /// Last 4 digits of the credit card used for billing.
        /// </summary>
        public string CardLast4 { get; set; }

        /// <summary>
        /// Date the organization first subscribed to a paid plan.
        /// </summary>
        public DateTime? SubscribeDate { get; set; }

        /// <summary>
        /// Date the billing information was last changed.
        /// </summary>
        public DateTime? BillingChangeDate { get; set; }

        /// <summary>
        /// User id that the billing information was last changed by.
        /// </summary>
        public string BillingChangedByUserId { get; set; }

        /// <summary>
        /// Organization's current billing status.
        /// </summary>
        public BillingStatus BillingStatus { get; set; }

        /// <summary>
        /// The price of the plan that this organization is currently on.
        /// </summary>
        public decimal BillingPrice { get; set; }

        /// <summary>
        /// Maximum number of error occurrences allowed per month.
        /// </summary>
        public int MaxErrorsPerMonth { get; set; }

        /// <summary>
        /// Number of days stats data is retained.
        /// </summary>
        public int RetentionDays { get; set; }

        /// <summary>
        /// If true, the account is suspended and can't be used.
        /// </summary>
        public bool IsSuspended { get; set; }

        /// <summary>
        /// The code indicating why the account was suspended.
        /// </summary>
        public string SuspensionCode { get; set; }

        /// <summary>
        /// Any notes on why the account was suspended.
        /// </summary>
        public string SuspensionNotes { get; set; }

        /// <summary>
        /// The reason the account was suspended.
        /// </summary>
        public DateTime? SuspensionDate { get; set; }

        /// <summary>
        /// User id that the suspended the account.
        /// </summary>
        public string SuspendedByUserId { get; set; }

        /// <summary>
        /// If true, premium features will be enabled.
        /// </summary>
        public bool HasPremiumFeatures { get; set; }

        /// <summary>
        /// Maximum number of users allowed by the current plan.
        /// </summary>
        public long MaxUsers { get; set; }

        /// <summary>
        /// Maximum number of projects allowed by the current plan.
        /// </summary>
        public int MaxProjects { get; set; }

        /// <summary>
        /// Total number of projects.
        /// </summary>
        public int ProjectCount { get; set; }

        /// <summary>
        /// Current number of error stacks in the system.
        /// </summary>
        public long StackCount { get; set; }

        /// <summary>
        /// Current number of error occurrences in the system.
        /// </summary>
        public long ErrorCount { get; set; }

        /// <summary>
        /// The date that the latest error occurred.
        /// </summary>
        public DateTime LastErrorDate { get; set; }

        /// <summary>
        /// Total errors logged by our system.
        /// </summary>
        public long TotalErrorCount { get; set; }

        /// <summary>
        /// Organization invites.
        /// </summary>
        public ICollection<Invite> Invites { get; set; }

        /// <summary>
        /// Hours over error limit.
        /// </summary>
        public ICollection<UsageInfo> OverageHours { get; set; }

        /// <summary>
        /// Account error usage information.
        /// </summary>
        public ICollection<UsageInfo> Usage { get; set; }
    }

    public class UsageInfo {
        public DateTime Date { get; set; }
        public int Total { get; set; }
        public int Accepted { get; set; }
        public int Limit { get; set; }
    }

    public enum BillingStatus {
        Trialing = 0,
        Active = 1,
        PastDue = 2,
        Canceled = 3,
        Unpaid = 4
    }
}